client/client.go

```diff
@@ -12,63 +12,118 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-// Package client contains generic representations of clients connecting to different receivers
+/*
+Package client contains generic representations of clients connecting to different receivers. Components,
+such as processors or exporters, can make use of this information to make decisions related to
+grouping of batches, tenancy, load balancing, tagging, among others.
+
+The structs defined here are typically used within the context that is propagated down the pipeline,
+with the values being produced by authenticators or receivers, and consumed by processors and exporters.
+
+Producers
+
+Authenticators are responsible for obtaining a client.Info from the current context and enhancing the client.Info
+with an implementation of client.AuthData. The attribute names should be documented with their return types and
+considered part of the public API for the authenticator.
+
+Receivers are responsible for obtaining a client.Info from the current context and enhancing the client.Info
+with the net.Addr from the peer. For HTTP requests, this is typically the IP address of the client. This new context
+should then be propagated down the pipeline when calling the next consumer. Typically, however, receivers would delegate
+this processing to helpers such as the confighttp or configgrpc packages: both contain interceptors that will enhance
+the context with the client.Info, such that no actions are needed by receivers that are built using confighttp.HTTPServerSettings
+or configgrpc.GRPCServerSettings.
+
+Consumers
+
+Processors and exporters have access to the client.Info via client.FromContext. Among other usages, this data can be used to:
+
+- annotate data points with authentication data (username, tenant, ...)
+
+- route data points based on authentication data
+
+- rate limit client calls based on IP addresses
+
+Neither processors nor exporters are expected to change data from the client.Info contained in the context, although
```

Comment:
Mutation should be prohibited or we must clone the client.Info in the fanout attached to multiple pipelines otherwise we allow concurrent modification.